pkg/apiserver/registry/system/supportbundle/rest.go

```diff
@@ -0,0 +1,373 @@
+	}
+	r.statusLocker.Lock()
+	defer r.statusLocker.Unlock()
+	if r.cancelFunc != nil {
+		r.cancelFunc()
+	}
+	r.cache = &supportBundle{
+		SupportBundle: systemv1beta1.SupportBundle{
+			ObjectMeta: metav1.ObjectMeta{Name: r.mode},
+			Status:     systemv1beta1.SupportBundleStatusNone,
+		},
+	}
+	return nil, true, nil
+}
+
+func (r *supportBundleREST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *supportBundleREST) collect(ctx context.Context, dumpers ...func(string) error) (*supportBundle, error) {
+	basedir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+	if err != nil {
+		return nil, fmt.Errorf("error when creating tempdir: %w", err)
+	}
+	defer defaultFS.RemoveAll(basedir)
+	for _, dumper := range dumpers {
+		if err := dumper(basedir); err != nil {
+			return nil, err
+		}
+	}
+	outputFile, err := defaultFS.Create(filepath.Join(afero.GetTempDir(defaultFS, ""), fmt.Sprintf("bundle_%d.tar.gz", rand.Int())))
+	if err != nil {
+		return nil, fmt.Errorf("error when creating output tarfile: %w", err)
+	}
+	defer outputFile.Close()
+	hashSum, err := packDir(basedir, outputFile)
+	if err != nil {
+		return nil, fmt.Errorf("error when packaing supportBundle: %w", err)
+	}
+
+	select {
+	case <-ctx.Done():
+		_ = defaultFS.Remove(outputFile.Name())
+		return nil, fmt.Errorf("collecting is canceled")
+	default:
+	}
+	stat, err := outputFile.Stat()
+	var fileSize int64
+	if err == nil {
+		fileSize = stat.Size()
+	}
+	creationTime := metav1.Now()
+	deletionTime := metav1.NewTime(creationTime.Add(bundleExpireDuration))
+	return &supportBundle{
+		SupportBundle: systemv1beta1.SupportBundle{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:              r.mode,
+				CreationTimestamp: creationTime,
+				DeletionTimestamp: &deletionTime,
+			},
+			Status: systemv1beta1.SupportBundleStatusCollected,
+			Sum:    fmt.Sprintf("%x", hashSum),
+			Size:   uint32(fileSize),
+		},
+		filepath: outputFile.Name(),
+	}, nil
+}
+
+func (r *supportBundleREST) collectAgent(ctx context.Context) (*supportBundle, error) {
+	dumper := support.NewAgentDumper(defaultFS, defaultExecutor, r.ovsCtlClient)
+	return r.collect(
+		ctx,
+		dumper.DumpLog,
+		dumper.DumpHostNetworkInfo,
+		dumper.DumpFlows,
+		dumper.DumpNetworkPolicyResources,
+		dumper.DumpAgentInfo,
+	)
+}
+
+func (r *supportBundleREST) collectController(ctx context.Context) (*supportBundle, error) {
+	dumper := support.NewControllerDumper(defaultFS, defaultExecutor)
+	return r.collect(
+		ctx,
+		dumper.DumpLog,
+		dumper.DumpNetworkPolicyResources,
+		dumper.DumpControllerInfo,
+	)
+}
+
+func (r *supportBundleREST) clean(ctx context.Context, bundlePath string, duration time.Duration) {
+	select {
+	case <-ctx.Done():
+	case <-time.After(duration):
+		func() {
+			r.statusLocker.Lock()
+			defer r.statusLocker.Unlock()
+			select { // check the context again in case of cancellation when acquiring the lock.
+			case <-ctx.Done():
```

Comment:
why this is needed when there's a default?

Reply:
Considering conditions:
- Running into this select
- A new bundle had just generated in another goroutine
We should not reset the status of the newly generated support bundle but only remove the old support bundles.

Reply:
So the "cancellation" means cancellation of cleanup? but it actually means a new create request?

Reply:
Cancellation cancels the context, in other words, tasks under the context are done.

Reply:
Got it, thanks for the explanation.